diff --git a/node_modules/react-native-vision-camera/ios/CameraView.swift b/node_modules/react-native-vision-camera/ios/CameraView.swift
index 8dd947e..74b2780 100644
--- a/node_modules/react-native-vision-camera/ios/CameraView.swift
+++ b/node_modules/react-native-vision-camera/ios/CameraView.swift
@@ -75,6 +75,7 @@ public final class CameraView: UIView, CameraSessionDelegate {
   }
 
   // pragma MARK: Internal Properties
+  var cameraViewController: CameraViewController?
   var cameraSession: CameraSession
   var isMounted = false
   var isReady = false
@@ -120,8 +121,16 @@ public final class CameraView: UIView, CameraSessionDelegate {
   }
   
   override public func layoutSubviews() {
-    previewView.frame = frame
-    previewView.bounds = bounds
+    super.layoutSubviews()
+    
+    // To disable jank double animation from react-native and the preview subview both rotating
+    CATransaction.begin()
+    CATransaction.setDisableActions(true)
+    layer.sublayers?.forEach({ layer in
+      layer.frame = frame
+      layer.bounds = bounds
+    })
+    CATransaction.commit()
   }
 
   func getPixelFormat() -> PixelFormat {
@@ -251,6 +260,8 @@ public final class CameraView: UIView, CameraSessionDelegate {
       }
     }
     
+    self.setupViewController()
+
     // Prevent phone from going to sleep
     UIApplication.shared.isIdleTimerDisabled = isActive
   }
@@ -269,6 +280,28 @@ public final class CameraView: UIView, CameraSessionDelegate {
     #endif
   }
   
+  // pragma MARK: CameraViewController
+  // This exists purely for orientation handling
+  
+  func setupViewController() {
+      if cameraViewController == nil {
+        cameraViewController = CameraViewController()
+        cameraViewController!.parentView = self
+          
+          DispatchQueue.main.async { [self] in
+              if self.reactViewController() != nil {
+                  self.reactViewController()!.addChild(self.cameraViewController!)
+                  self.cameraViewController!.didMove(toParent: self.reactViewController())
+              } else {
+                  self.reactAddController(toClosestParent: self.cameraViewController)
+              }
+          }
+          
+          cameraViewController!.view.frame = self.frame
+          self.addSubview(cameraViewController!.view)
+      }
+  }
+
   // pragma MARK: Event Invokers
 
   func onError(_ error: CameraError) {
@@ -353,18 +386,20 @@ public final class CameraView: UIView, CameraSessionDelegate {
     guard let cameraPosition = cameraSession.videoDeviceInput?.device.position else {
       return .up
     }
-    let orientation = cameraSession.configuration?.orientation ?? .portrait
+    let orientation = previewView.videoPreviewLayer.connection?.videoOrientation
 
-    // TODO: I think this is wrong.
+    // TODO: I think this is right...
     switch orientation {
     case .portrait:
-      return cameraPosition == .front ? .leftMirrored : .right
+      return cameraPosition == .front ? .downMirrored : .down
     case .landscapeLeft:
-      return cameraPosition == .front ? .downMirrored : .up
+      return cameraPosition == .front ? .leftMirrored : .left
     case .portraitUpsideDown:
-      return cameraPosition == .front ? .rightMirrored : .left
+      return cameraPosition == .front ? .upMirrored : .up
     case .landscapeRight:
-      return cameraPosition == .front ? .upMirrored : .down
+      return cameraPosition == .front ? .rightMirrored : .right
+    default:
+      return .up
     }
   }
 }
diff --git a/node_modules/react-native-vision-camera/ios/CameraViewController.swift b/node_modules/react-native-vision-camera/ios/CameraViewController.swift
new file mode 100644
index 0000000..20425c0
--- /dev/null
+++ b/node_modules/react-native-vision-camera/ios/CameraViewController.swift
@@ -0,0 +1,49 @@
+//
+//  CameraViewController.swift
+//  VisionCamera
+//
+//  Created by Max Gurela on 2/8/24.
+//
+
+import AVFoundation
+import UIKit
+
+public final class CameraViewController: UIViewController {
+  var parentView:CameraView!
+
+  public func overrideOrientation() {
+    if let videoPreviewLayerConnection = parentView.previewView.videoPreviewLayer.connection {
+      // Kinda jank, but this is the best way I could find to get the interface orientation in Swift 5/iOS 17..
+      guard let interfaceOrientation = UIApplication.shared.windows.first(where: { $0.isKeyWindow })?.windowScene?.interfaceOrientation else {
+        return
+      }
+      
+      // UIInterfaceOrientation overlaps perfectly with AVCaptureVideoOrientation
+      guard let newVideoOrientation = AVCaptureVideoOrientation(rawValue: interfaceOrientation.rawValue),
+            interfaceOrientation.isPortrait || interfaceOrientation.isLandscape else {
+        return
+      }
+      
+      videoPreviewLayerConnection.videoOrientation = newVideoOrientation
+    }
+  }
+  
+  public override func viewWillLayoutSubviews() {
+    super.viewDidLayoutSubviews()
+    
+    self.overrideOrientation()
+  }
+ 
+  public override func viewDidAppear(_ animated: Bool) {
+    super.viewDidAppear(animated)
+    
+    self.overrideOrientation()
+  }
+  
+  public override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
+    super.viewWillTransition(to: size, with: coordinator)
+    
+    self.overrideOrientation()
+  }
+  
+}
diff --git a/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureOutput+mirror.swift b/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureOutput+mirror.swift
index 63910cd..1b1b05a 100644
--- a/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureOutput+mirror.swift
+++ b/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureOutput+mirror.swift
@@ -30,6 +30,8 @@ extension AVCaptureOutput {
    - For Videos, the buffers are physically rotated if available, since we use an AVCaptureVideoDataOutput instead of an AVCaptureMovieFileOutput.
    */
   func setOrientation(_ orientation: Orientation) {
+    return;
+    
     // Set orientation for each connection
     for connection in connections {
       #if swift(>=5.9)
diff --git a/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureVideoOrientation+initWithInterfaceOrientation.swift b/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureVideoOrientation+initWithInterfaceOrientation.swift
new file mode 100644
index 0000000..ee3f167
--- /dev/null
+++ b/node_modules/react-native-vision-camera/ios/Extensions/AVCaptureVideoOrientation+initWithInterfaceOrientation.swift
@@ -0,0 +1,21 @@
+//
+//  AVCaptureVideoOrientation+initWithInterfaceOrientation.swift
+//  VisionCamera
+//
+//  Created by Max Gurela on 4/2/24.
+//
+
+import AVFoundation
+
+extension AVCaptureVideoOrientation {
+  init(ui:UIInterfaceOrientation) {
+    // AVCaptureVideoOrientation has left/right swapped from the interface orientation enum... Thanks Apple!
+    switch ui {
+      case .landscapeRight:       self = .landscapeRight
+      case .landscapeLeft:        self = .landscapeLeft
+      case .portrait:             self = .portrait
+      case .portraitUpsideDown:   self = .portraitUpsideDown
+      default:                    self = .portrait
+    }
+  }
+}
